EB-618: Add initial DivBase architectural decision record (ADR) Proposal (#18)
ScilifelabDataCentre/divbase: adr/002-API-design.md

# 1. ADR 002: Initial Design of DivBase API

## 2. Status
Proposed

## 3. Context
DivBase API needs to handle all user interactions with the system:

> - users/projects management.  

Reviewer: ```suggestion
- User and project management.  
```

- Facilitate upload/download of data files to S3 (users will not have accounts with S3).
- Run queries on the datasets and associated metadata "results files".

**Considerations:**
- The API must support both web-based and CLI-based workflows.
- Users will interact with the object store indirectly via pre-signed URLs generated by the API.
- We are making a CLI tool which handles interacting with the API. Whilst users could use the API directly, we are not optimizing/aiming for that.  

> - The API will send a lot of network requests (to/from: user, job managenemnt system and S3) so an asynchronous implementation would be ideal

Reviewer: ```suggestion
- The API will send a lot of network requests (to/from: user, job management system and S3) so an asynchronous implementation would be ideal.
```

The following endpoints will be implemented:

FRONTEND_URLs:
```bash
{register, login, logout, reset_password}/
project/{proj_id} Basic overview, see projects, manage members if project manager
users/: Edit name, email, change password etc...
```

API_URLs:
```bash
/api/v1/{login, logout}/
/api/v1/query/: Submit query or check the status of a query
/api/v1/files/: Create presigned URL for user for S3 upload/download
/api/v1/version/: Utility to version the state of all bucket files
```

### Auth Flows

Propose to use JSON Web Tokens (JWT) as well supported by FastAPI. Can reuse the same auth logic as much as possible for both the Frontend and API.  

<p align="center">
  <img src="imgs/002-web-auth-flow.png" alt="Web auth flow" width="800">
  <br>
  <em>Auth flow for a web browser interacting with DivBase API. (API Router and API Token Service are both part of the FastAPI app).</em>
</p>

<p align="center">
  <img src="imgs/002-jwt-refresh.png" alt="JWT web refresh token flow" width="800">
  <br>
  <em>Refresh token logic. Unlike the access token, the refresh token logic includes a db lookup to verify that the user is still active and not deleted.</em>
</p>

> ### CLI vs webbrowser Difference:

Reviewer: ```suggestion
### CLI vs Web Browser Difference:
```

> The webbrowser would store the tokens in HttpOnly Cookies. 

Reviewer: ```suggestion
The web browser would store the tokens in HttpOnly Cookies. 
```

> The CLI would store the access and refresh tokens on the users local pc: E.g.:

Reviewer: ```suggestion
The CLI would store the access and refresh tokens on the users local PC, e.g.:
```

```bash
divbase-cli auth login ...

> ## tokens auto added to ~/.config/.divbase_cli/.env

Reviewer 2: I salute to that

divbase-cli query ... # Both tokens passed in HTTP request sent to the DivBase API. 
```

### DB Models - defined with sqlalchemy 2
```bash
Base: # inherited by all other models
  UUID, created_at, updated_at

Users(Base): 
  name, email, hashed_password, is_admin, is_active

Projects (Base): 
    name, description, storage_quota, bucket_name, is_active

ProjectMembership(Base): # To handle many-to-many relationship between users and projects
    user_id, project_id, role*, joined_at, is_active
    # *role would be an enum: {READ, WRITE, MANAGE}

# optional 
- RevokedRefreshTokens(Base) # For if users logs out or want could have this as a blacklist table. 
```

**Why no Jobs Table?** Our current approach is that the DivBase API queries the FlowerAPI for the job status/details and passes that back to the user. (see diagram in [ADR-001](adr/imgs/001-initial-design.png))

## 5. Consequences
Positive:
- FastAPI provides native async support. 
- JWTs can be used by both browser and API (reduced complexity).  
- Pre-signed URLs mean only once place for user administration. 

Negative:

> - JWTs being stateless Token revocation requires additional infrastructure. That said users will only have soft delete powers on S3 (so a leaked access/refresh token cannot do too much damage). 

Reviewer 3: ## Token Revocation Strategy ✅

**Your approach looks great!** Using `last_password_change` + `is_active` checks is a smart pattern that avoids storing all active refresh tokens.

### What You've Implemented

Your flow is solid:
- **Explicit revocation:** `RevokedRefreshTokens` table for logout/manual revoke
- **Implicit revocation:** Check `refresh_token.iat < user.last_password_change` 
- **Account status:** Check `user.is_active`

This is actually **better** than storing all refresh tokens because:
- ✅ No storage overhead for every token issued
- ✅ Password change automatically invalidates all sessions (good security)
- ✅ DB lookup only on refresh (not every request)
- ✅ Simpler cleanup (only explicit revocations need TTL management)

### One Clarification

In your flow, you wrote:
> "when a user hits the refresh endpoint, the Users db model would be used to check if refresh token `iat` is older than `user.last_password_change`"

**This is correct!** The `iat` (issued at) claim is in the JWT itself, so you can:
1. Decode the refresh token (don't need to verify signature yet for this check)
2. Compare `token.iat < user.last_password_change`
3. If older → reject immediately (password changed since token issued)
4. If newer → proceed with full verification

**Alternative phrasing for ADR (optional):**
```markdown
**On refresh request:**
1. Check if token JTI in RevokedRefreshTokens → reject if found
2. Fetch user from DB by user_id (from token)
3. Check user.is_active → reject if False
4. Check token.iat < user.last_password_change → reject if older
5. If all checks pass → issue new tokens
```

### Minor Enhancement Suggestion

Consider adding `last_login` timestamp to Users table:
```python
Users(Base):
    ...
    last_password_change = Column(DateTime)
    last_login = Column(DateTime)  # Update on successful refresh
```

**Benefits:**
- Can identify inactive accounts (haven't logged in for 6+ months)
- Security monitoring (detect compromised accounts with unusual login times)
- User analytics (active vs dormant users)

**Not required, just nice-to-have.**

---

**Bottom line:** Your token revocation strategy is production-ready. This blocker is resolved. ✅

- Dependency on FlowerAPI for job management introduces potential latency.

## 6. Alternatives Considered:

> - Django/Flask: Not arguing one is better than the other. Due to the low priority of having a frontend and that the frontend is expected to be quite limited, we have decided to try out FastAPI initially. The admin panel that comes with Django would be one benefit of using Django instead (make viewing users/projects easier). The support for async in FastAPI would be one advantage of using FastAPI.    

Reviewer 2: Other than already mentioned django-ninja, I should also mention that django has async support, but I'm not experienced enough here. Nonetheless, it would be interesting to check where this set would work for you.

Author: Django seems to have partial async support (e.g. not for the db), but I doubt this would be truly limiting for our application. For the admin panel I'm going to test out either https://github.com/jowilf/starlette-admin or https://github.com/aminalaee/sqladmin in order to get an equivalent to Django admin with hopefully little work but for FastAPI.
